contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,262 @@
+        _bnt = bnt;
+    }
+
+    /**
+     * @dev fully initializes the contract and its parents
+     */
+    function initialize() external initializer {
+        __CarbonVortex_init();
+    }
+
+    // solhint-disable func-name-mixedcase
+
+    /**
+     * @dev initializes the contract and its parents
+     */
+    function __CarbonVortex_init() internal onlyInitializing {
+        __Upgradeable_init();
+        __ReentrancyGuard_init();
+
+        __CarbonVortex_init_unchained();
+    }
+
+    /**
+     * @dev performs contract-specific initialization
+     */
+    function __CarbonVortex_init_unchained() internal onlyInitializing {
+        setRewards(100_000);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 2;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function setRewards(uint256 newRewardsPercentagePPM) public onlyAdmin validFee(uint32(newRewardsPercentagePPM)) {
+        uint256 prevRewardsPercentagePPM = _rewardsPercentagePPM;
+
+        // return if the rewards percentage PPM is the same
+        if (prevRewardsPercentagePPM == newRewardsPercentagePPM) {
+            return;
+        }
+
+        _rewardsPercentagePPM = newRewardsPercentagePPM;
+
+        emit RewardsUpdated({ prevRewards: prevRewardsPercentagePPM, newRewards: newRewardsPercentagePPM });
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function rewards() external view returns (uint256) {
+        return _rewardsPercentagePPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function totalBurnt() external view returns (uint256) {
+        return _totalBurnt;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function availableFees(Token token) external view returns (uint256) {
+        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], _rewardsPercentagePPM, PPM_RESOLUTION);
```

Comment:
Cache `_rewardsPercentagePPM` in order to avoid reading a storage slot on every iteration

Reply:
cached